### PR TITLE
Support creation of build and deployment secrets

### PR DESCRIPTION
**Pull request by karthikjeeyar**
## Fixes 

https://issues.redhat.com/browse/HAC-4560
https://issues.redhat.com/browse/HAC-4913

## Description

Adds support to create build and deployment secret.

## Type of change

- [x] Feature

## Screen shots / Gifs for design review 

**Empty state & Learn more panel:** 

image

**Validation error messages:**
---
Screenshot 2023-09-18 at 7 40 33 PM 

Invalid json:
Screenshot 2023-09-20 at 4 00 28 PM
---

**Build Secret form:**

image

**ImagePull secret (.dockerconfigjson) from username/password form:**

image

**Deployment Secret form:** 

image

**Secret form in action:**

https://github.com/openshift/hac-dev/assets/9964343/172b4870-e97e-4872-be50-816b0483c7a2

## Browser conformance: 

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge

cc: @pankajshivpuje @MisJohns

### CODE REVIEW

File: src/components/Secrets/SecretsForm/EncodedFileUploadField.tsx [hunk 1]
```diff
@@ -19,17 +22,26 @@ const EncodedFileUploadField: React.FC<EncodedFileUploadFieldProps> = ({
...
   const [filename, setFilename] = React.useState<string>();
-  const [, { value }, { setValue }] = useField<string>(name);
+  const [, { value, error, touched }, { setValue, setTouched }] = useField<string>(name);
+  const isValid = !(touched && error);
 
   const onChange = React.useCallback(
-    (data: string) => {
-      setValue(data ? Base64.encode(data) : '', true);
+    (data: string, fileUploaded?: boolean) => {
+      const parsedData = attempt(JSON.parse, data);
```
Comment (arewm) [inline src/components/Secrets/SecretsForm/EncodedFileUploadField.tsx:30]: Is this where the contents of the .dockerconfig are loaded to be put into the k8s secret? Users may not realize that they need to upload a JSON version of the secret. Would it be possible to also accept the base64 encoded data (i.e. that they may pull out of their own secret) to just be inserted without modification?

Or would this overly complicate the user flow, error detection, and user comprehension?

Comment (arewm) [reply src/components/Secrets/SecretsForm/EncodedFileUploadField.tsx:30]: @pankajshivpuje , fyi

Comment (karthikjeeyar) [reply src/components/Secrets/SecretsForm/EncodedFileUploadField.tsx:30]: > Is this where the contents of the .dockerconfig are loaded to be put into the k8s secret? Users may not realize that they need to upload a JSON version of the secret.

@arewm Yes, I have added a validation message to guide the user to enter the JSON data. This format is similar to how we add dockerconfigjson secret in Openshift UI.



image

> Or would this overly complicate the user flow, error detection, and user comprehension?

Right, handling both base64 value as well as JSON fortmat might become slightly complicated flow.

**Review by rottencandy — COMMENTED** [review]
For some reason, validation is not running on switching secret type, I tried testing with `validateOnChange` but it doesn't seem to fix it :/

https://github.com/openshift/hac-dev/assets/20013884/3699ff52-911d-48ff-a33d-071dbba3b546

**Comment by karthikjeeyar** [comment]
> For some reason, validation is not running on switching secret type, I tried testing with validateOnChange but it doesn't seem to fix it :/

Hmm.. I couldn't reproduce this, It disable/enable the create button while switching types 🤔.

**Review by rottencandy — APPROVED** [review]

**Pull request merged**